# Tutorial 3: Metagenomics Data Analysis

## Data preparation
Download the sample data from [here](https://drive.google.com/file/d/1_8hZF7uIzIIpglB3nXRYNzkuqoZTmKo-/view?usp=sharing) (43 MB).

**Note**: in a real metagenomics analysis, you should first check the sequencing data quality (e.g., with FastQC, see Tutorial 1), correct/trim it if needed (see Tutorial 1), and perform the host removal (see Tutorial 2). However, this sample data has already been preprocessed for you, so just unpack the tarball and proceed directly to the assembly step.

In [4]:
!pip install biopython

Defaulting to user installation because normal site-packages is not writeable


In [2]:
from Bio import SeqIO

In [12]:
read1 = list(SeqIO.parse("reads/ERR2231572_subsampled_1.fastq", "fastq"))
head(read1, n=5)

ID: ERR2231572.91
Sequence: GTTGATGAGGGTGGTGTGATTAGTCGACAACTTGGAACAGTTAAACCGAATGACAATTACGAGATTATGANCGGTTGGACGNTATCTAAGANNGNNNNATATGCANTTGAGTATGACAGCGACAAGTTAGTNGCCGAGTGGCACGTTTAGGAGAAAAATTATGATTAACACATTATTGATTANCGNGANTGTTTTTNNNNNNNTTTGGTTNNNNNNNCTCGTATNNNNNNNTACCTTTG
Quality: [34, 34, 34, 34, 34, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 38, 38, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 37, 38, 38, 37, 38, 37, 38, 38, 38, 2, 25, 34, 37, 37, 35, 38, 38, 38, 38, 38, 2, 25, 35, 37, 34, 37, 38, 38, 38, 38, 2, 2, 24, 2, 2, 2, 2, 25, 25, 34, 37, 38, 34, 37, 2, 25, 33, 33, 37, 38, 38, 38, 38, 37, 38, 38, 38, 38, 38, 38, 33, 37, 38, 35, 38, 30, 36, 37, 38, 38, 2, 25, 30, 30, 37, 38, 34, 31, 37, 38, 38, 37, 38, 38, 29, 37, 38, 38, 38, 38, 38, 37, 35, 37, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 28, 37, 38, 38, 38, 34, 32, 32, 38, 32, 38, 38, 38, 38, 32, 2, 1

In [13]:
read2 = list(SeqIO.parse("reads/ERR2231572_subsampled_2.fastq", "fastq"))
head(read2, n=5)

ID: ERR2231572.91
Sequence: NAAAGGTANNNNNNNATACGAGNNNNNNNAACCAAAGTNATCCAAAAACAATCGCGATNNTCNATAATGTGNNNATCATAATTTTTCTCCTAAACGTNCCACTCGGNNANNNNCTTGTCGCNGTCATACTCAAGTNCANNTAGTTCTTTCTTAGATATCGTCCAACCGTNNATAANCTCGTAATNGTNATTCGGTTTAACTGTTCNNNGTTGTCGACTNNTNNNNCCACCCTCATCAA
Quality: [2, 23, 32, 34, 34, 38, 38, 38, 2, 2, 2, 2, 2, 2, 2, 28, 28, 35, 37, 38, 38, 38, 2, 2, 2, 2, 2, 2, 2, 25, 25, 35, 37, 37, 38, 38, 38, 37, 2, 25, 35, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 2, 2, 25, 25, 2, 25, 34, 37, 38, 38, 37, 36, 38, 2, 2, 2, 25, 25, 34, 37, 35, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 2, 24, 33, 37, 38, 35, 36, 38, 38, 2, 2, 25, 2, 2, 2, 2, 24, 25, 24, 37, 38, 38, 38, 38, 2, 10, 24, 37, 38, 38, 38, 38, 37, 38, 38, 35, 34, 25, 2, 23, 23, 2, 2, 23, 23, 32, 37, 35, 37, 38, 37, 38, 38, 38, 38, 38, 38, 35, 38, 38, 37, 38, 38, 38, 37, 38, 38, 38, 38, 38, 38, 34, 2, 2, 10, 21, 31, 29, 2, 19, 20, 35, 34, 37, 36, 34, 37, 2, 9, 20, 2, 19, 19, 31

## MetaSPAdes

### Install SPAdes Toolkit

In [15]:
!conda install -c bioconda spades -y

Solving environment: done

## Package Plan ##

  environment location: /Users/Erfan/miniconda3

  added / updated specs:
    - spades


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    spades-4.0.0               |       h07b6146_3        13.5 MB  bioconda
    ------------------------------------------------------------
                                           Total:        13.5 MB

The following NEW packages will be INSTALLED:

  spades             bioconda/osx-arm64::spades-4.0.0-h07b6146_3 



                                                                                
Preparing transaction: done
Verifying transaction: done
Executing transaction: -     
#############################################################################################
#                                                                                           #
#   Note: SPAdes installed through bioconda

### Run MetaSPAdes with Default Settings:

In [16]:
!metaspades.py --only-assembler -o output_default -1 reads/ERR2231572_subsampled_1.fastq -2 reads/ERR2231572_subsampled_2.fastq

Command line: /Users/Erfan/miniconda3/bin/metaspades.py	--only-assembler	-o	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/output_default	-1	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_1.fastq	-2	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_2.fastq	

System information:
  SPAdes version: 4.0.0
  Python version: 3.9.6
  OS: macOS-12.4-arm64-arm-64bit

Output dir: /Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/output_default
Mode: ONLY assembling (without read error correction)
Debug mode is turned OFF

Dataset parameters:
  Metagenomic mode
  Reads:
    Library number: 1, library type: paired-end
      orientation: fr
      left reads: ['/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_1.fastq']
      right reads: ['/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_2.fastq']
      interlaced reads: not specified
     

### Run MetaSPAdes with K-mer Length Set to 21:

In [17]:
!metaspades.py --only-assembler -k 21 -o output_k21 -1 reads/ERR2231572_subsampled_1.fastq -2 reads/ERR2231572_subsampled_2.fastq

Command line: /Users/Erfan/miniconda3/bin/metaspades.py	--only-assembler	-k	21	-o	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/output_k21	-1	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_1.fastq	-2	/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_2.fastq	

System information:
  SPAdes version: 4.0.0
  Python version: 3.9.6
  OS: macOS-12.4-arm64-arm-64bit

Output dir: /Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/output_k21
Mode: ONLY assembling (without read error correction)
Debug mode is turned OFF

Dataset parameters:
  Metagenomic mode
  Reads:
    Library number: 1, library type: paired-end
      orientation: fr
      left reads: ['/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_1.fastq']
      right reads: ['/Users/Erfan/Documents/GitHub/Metagenomics_Data_Analysis/reads/ERR2231572_subsampled_2.fastq']
      interlaced reads: not specified
      s

### Copy and Rename the contigs.fasta Files:

In [18]:
# copy the contigs.fasta file to the current directory
!cp output_default/contigs.fasta metaspades_default.fasta

In [19]:
# copy the contigs.fasta file to the current directory
!cp output_k21/contigs.fasta metaspades_k21.fasta

In [20]:
# remove the output directories for cleanliness and to save space
!rm -rf output_default output_k21

## MEGAHIT

In [28]:
## create a new environment
!CONDA_SUBDIR=osx-64 conda create -n Microbiome_env -c conda-forge -c bioconda megahit -y

Solving environment: done

## Package Plan ##

  environment location: /Users/Erfan/miniconda3/envs/Microbiome_env

  added / updated specs:
    - megahit


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2016.9.26          |           py36_0         217 KB  conda-forge
    libcxx-19.1.4              |       hf95d169_0         517 KB  conda-forge
    libffi-3.3                 |       h046ec9c_2          45 KB  conda-forge
    llvm-openmp-19.1.4         |       ha54dae1_0         298 KB  conda-forge
    megahit-1.2.9              |       hfbae3c0_0         1.4 MB  bioconda
    openssl-1.1.1w             |       h8a1eda9_0         1.7 MB  conda-forge
    pip-20.0.2                 |           py36_1         1.9 MB  conda-forge
    python-3.6.13              |       h88f2d9e_0        16.8 MB
    python_abi-3.6             |          2_cp36m           4 KB  conda-forge
    setuptool

<div class="alert alert-block alert-info">
<b>Activate Environment:</b> 

`$conda activate Microbiome_env`

</div>

MEGAHIT's official documentation is [here](https://github.com/voutcn/megahit/blob/master/README.md). We will use v.1.2.9 ([in conda](https://anaconda.org/bioconda/megahit)).

Run the tool on our data only once with the **K-mer length set to 21**. Please use the following memory-saving option: `--mem-flag 0`.

MEGAHIT produces some extra files in the output directory. We will use only `final.contigs.fa`, so copy it with a meaningful name (e.g.,  `megahit_k21.fasta` ) and remove the remaining output directory to free up space.

<div class="alert alert-block alert-info">
<b> MEGAHIT: </b> 

`$megahit -1 reads/ERR2231572_subsampled_1.fastq -2 reads/ERR2231572_subsampled_2.fastq -o output_directory --k-list 21 --mem-flag 0`

</div>

### Metagenome assembly evaluation: part 1


We will assess the quality of the metagenome assemblies with [QUAST](https://quast.sf.net/).

Install QUAST **v.5.3.0** (e.g., from [conda](https://anaconda.org/bioconda/quast)), run it with all three obtained assemblies together (two from metaSPAdes and one from MEGAHIT), and check out the final quality report.

 

What is the **highest N50** value among the assemblies?

Python 3.8, 3.9, 3.10, 3.11, or 3.12, but Python 3.7 or lower, and Python 3.13 or higher, are not compatible with this version of QUAST. So we create a new Environment.

<div class="alert alert-block alert-info">
<b>Activate Environment:</b> 


`$conda create -n Microbiome_env_quast python=3.12`

`$conda activate Microbiome_env_quast`

`$conda install -c bioconda quast=5.3.0`

</div>

<div class="alert alert-block alert-info">
<b> Installing Quast: </b> 

`$conda install -c bioconda quast=5.3.0`

</div>

After the installation, we can run QUAST to analyze the three assemblies we have (two from metaSPAdes and one from MEGAHIT) by specifying the paths to these assembly files.

<div class="alert alert-block alert-info">
<b> Analyse Quast: </b> 

`$quast.py -o quast_output reads/ERR2231572_subsampled_1.fastq reads/ERR2231572_subsampled_2.fastq output_directory/final.contigs.fa`

</div>

**Check the final quality report:**
After running the command, QUAST will generate an output directory (quast_output). Inside this directory, there will be a file named report.html which we can open in a browser to review the quality metrics.

In the generated report, look for the N50 value under the summary section for each assembly. 